packages/data-broker/src/services/DataServiceResolver.js

```diff
@@ -0,0 +1,181 @@
+   * which stores data for multiple countries against that data element. If you ask for the
+   * data service for that data element, you should get the single DHIS Server.
+   *
+   * @public
+   * @param {DataSource[]} dataSources
+   * @param {Entity[]} [orgUnits]
+   * @returns {Promise<DataServiceMapping>}
+   */
+  async getMappingMultipleOrgUnits(dataSources, orgUnits = []) {
+    const dataElements = dataSources.filter(ds => ds.databaseType === TYPES.DATA_ELEMENT);
+    const dataGroups = dataSources.filter(ds => ds.databaseType === TYPES.DATA_GROUP);
+
+    const mapping = new DataServiceMapping();
+    mapping.dataElementMapping = await this.resolveDataElements(dataElements, orgUnits);
+    mapping.dataGroupMapping = await this.resolveDataGroups(dataGroups);
+    return mapping;
+  }
+
+  /**
+   * Convenience method
+   * @see getMapping
+   * @public
+   * @param {DataSource[]} dataSources
+   * @param {string} [orgUnitCode]
+   * @returns {Promise<DataServiceMapping>}
+   */
+  async getMappingByOrgUnitCode(dataSources, orgUnitCode) {
+    const orgUnit = orgUnitCode
+      ? await this.models.entity.findOne({ code: orgUnitCode })
+      : undefined;
+    return this.getMapping(dataSources, orgUnit);
+  }
+
+  /**
+   * Convenience method
+   * @see getMappingMultipleOrgUnits
+   * @public
+   * @param {DataSource[]} dataSources
+   * @param {string[]} orgUnitCodes
+   * @return {Promise<DataServiceMapping>}
+   */
+  async getMappingByOrgUnitCodes(dataSources, orgUnitCodes) {
+    if (Array.isArray(orgUnitCodes) && orgUnitCodes.length > 0) {
+      const orgUnits = await this.models.entity.find({ code: orgUnitCodes });
+      return this.getMappingMultipleOrgUnits(dataSources, orgUnits);
+    }
+    return this.getMapping(dataSources, undefined);
+  }
+
+  /**
+   * @private
+   * @param {DataElement[]} dataElement
+   * @param {Entity[]} [orgUnits]
+   * @returns {Promise<DataServiceMappingEntry[]>}
+   */
+  async resolveDataElements(dataElements, orgUnits = []) {
+    const countryCodes = orgUnits
+      .map(orgUnit => orgUnit.country_code)
+      .filter(countryCode => countryCode !== null && countryCode !== undefined);
+    const dataElementCodes = dataElements.map(de => de.code);
+
+    const mappings =
+      countryCodes.length > 0
+        ? await this.models.dataElementDataService.find({
+            country_code: countryCodes,
+            data_element_code: dataElementCodes,
+          })
+        : [];
+
+    const resolved = [];
+    for (const dataElement of dataElements) {
+      const deMappings = mappings.filter(m => m.data_element_code === dataElement.code);
+
+      if (deMappings.length > 1) {
+        // If they all resolve to the same data service, this is fine
+        const resolveToTheSameDataService = (mappingA, mappingB) => {
+          if (mappingA.service_type !== mappingB.service_type) return false;
+          return isequal(mappingA.service_config, mappingB.service_config);
+        };
```

Comment:
Up to you, but might be nice to move this function out to somewhere external as it doesn't require any local state 👍

Reply:
Reworked